stdlib/mlang/language-composer.mc

```diff
@@ -0,0 +1,355 @@
+              else None ()) decls in 
+    let synOrSemStrings = map nameGetStr synOrSemNames in 
+
+    match addImplicitIncludes (nameGetStr l.ident) includes synOrSemStrings ctx 
+    with (ctx, generatedDecls) in 
+
+    (ctx, DeclLang {l with decls = concat decls generatedDecls})
+  | other -> (ctx, other)
+
+  sem handleDecl : String -> [String] -> ComposerContext -> Decl -> (ComposerContext, Decl)
+  sem handleDecl langStr includes ctx = 
+  | decl & DeclSem d ->
+    let identStr = nameGetStr d.ident in 
+    let findMatchingInfo : String -> Option DeclInfo = lam incl.
+      mapLookup (incl, identStr) ctx.langMap in 
+    let foundIncludes : [DeclInfo] = mapOption findMatchingInfo includes in 
+    
+    let conflicts = filter (lam i. or (isTypeInfo i) (isSynInfo i)) foundIncludes in 
+    let errors = cons (d.info, nameGetStr d.ident) (map extractInfoName conflicts) in 
+
+    if not (null conflicts) then
+      errorMulti errors "The declared sem has an identifier that conflicts with included types!"
+    else
+      let includedSems = filter isSemInfo foundIncludes in 
+
+      let includes = map projIdent includedSems in 
+      (ctxWithDeclInfo ctx (langStr, nameGetStr d.ident) (decl2info langStr decl), 
+       DeclSem {d with includes = includes})
+  | decl & DeclSyn d ->
+    let identStr = nameGetStr d.ident in 
+    let findMatchingInfo : String -> Option DeclInfo = lam incl.
+      mapLookup (incl, identStr) ctx.langMap in 
+    let foundIncludes : [DeclInfo] = mapOption findMatchingInfo includes in 
+    
+    let conflicts = filter isTypeInfo foundIncludes in 
+
+    let errors = cons (d.info, nameGetStr d.ident) (map extractInfoName conflicts) in 
+
+    if not (null conflicts) then
+      errorMulti errors "The declared syn has an identifier that conflicts with included types!"
+    else
+      let includedSyns = filter isSynInfo foundIncludes in 
+
+      let includes = map projIdent includedSyns in 
+      let info = {ident = d.ident, info = d.info} in 
+      (ctxWithDeclInfo ctx (langStr, nameGetStr d.ident) (decl2info langStr decl), 
+       DeclSyn {d with includes = includes})
+  | decl & DeclType d -> 
+    let identStr = nameGetStr d.ident in 
+    let findMatchingInfo : String -> Option DeclInfo = lam incl.
+      mapLookup (incl, identStr) ctx.langMap in 
+    let conflicts : [DeclInfo] = mapOption findMatchingInfo includes in 
+
+    let errors = cons (d.info, nameGetStr d.ident) (map extractInfoName conflicts) in 
+
+    if not (null conflicts) then
+      errorMulti errors "The declared type has an identifier that conflicts with included types or syns!"
+    else
+      let info = decl2info langStr decl in 
+      (ctxWithDeclInfo ctx (langStr, nameGetStr d.ident) info, decl)
+  | _ -> error "Only Type, Syn, and Sem declarations can be contained inside of a langauge!"
+
+  sem addImplicitIncludes langStr includes definedSynsSems =
+  | ctx ->
+    -- We are going to remove elements from the ctx.langMap that
```

Comment:
I think there's something odd about the negations in this comment relative to the code, and generally double negations are hard.